src/providers/etherscan.js

```diff
@@ -1,29 +1,95 @@
+import rp from 'request-promise';
+import logger from '../logger';
+import { getAccessToken } from '../config';
+
+const path = require('path');
+
+const fs = (require('fs'));
+
+const expectedResponce = {
+  OK: 'OK',
+};
+
 /**
- * 
- * This is private function 
- * that saves the file ABI object into file locally
+ *  Validate if Etherscan response was succssefful should
+ *  throw if a reponse had an error
+ *  @param reponse: object
  *
- * @param address
- * return: string 
  */
-export const saveABIFile = data => 'FILE_PATH'; // eslint-disable-line no-unused-vars
-
+const ValidateResponce = (response) => {
+  if (response.status && response.message === expectedResponce.OK) {
+    return;
+  }
+  throw new Error('Wrong response from Etherscan or wrong Contract Address');
+};
 /**
- * 
- * This function will get the ABI from etherscan 
- * and pass it to saveABIFile function
+ *   Save Given ABI object in a local file
+ *   throw if a reponse had an error
+ *   @param abi: Object
+ *   @param filepath: string
  *
- * @param address
- * return: string 
+ */
+const saveABIFile = async (abi, filepath) => {
+  try {
+    fs.writeFileSync(filepath, abi);
+  } catch (err) {
+    throw new Error('Problem with writing ABI file');
+  }
+};
+/**
+ *   Scrape ABI from Etherscan
+ *   throw if a reponse had an error
+ *   @param address: string
+ *   @param jsonPath: string
+ *   @return object
+ */
+const ScrapeABI = async (address) => {
+  const options = {
+    uri: 'https://api.etherscan.io/api?module=contract',
+    qs: {
+      action: 'getabi',
+      address,
+      access_token: getAccessToken,
+    },
+    headers: {
+      'User-Agent': 'Request-Promise',
+    },
+    json: true,
+
+  };
+  try {
+    const abi = await rp(options);
+    ValidateResponce(await abi);
+    return JSON.parse(abi.result);
+  } catch (err) {
+    if (err === Error('Wrong response from Etherscan or wrong Contract Address')) { return err; }
+    return ScrapeABI(address);
```

Comment:
if the catch really executed thats may have possibility to never get out from this recursion.
Imagine execute this function as following
```
ScrapeABI('wront_address');
```